### PR TITLE
PyO3: Add `None` and `Tensor` indexing to `candle.Tensor`

### PR DESCRIPTION
**Pull request by LLukas22**
This adds the ability to expand tensor dimensions via a `None` index.
e.g.:
```python
t = candle.rand((12, 12))
c = t[:, None, None, :] # shape is now (12,1,1,12)
```

Tensor indexing is also supported but similarly to `candle-core` its limited to 1D tensors.

### CODE REVIEW

File: candle-pyo3/src/lib.rs [hunk 1]
```diff
@@ -475,27 +477,48 @@ impl PyTensor {
...
+                return Err(PyTypeError::new_err(
+                    "multi-dimensional tensor indexing is not supported",
+                ));
+            }
+            indexers.push(Indexer::IndexSelect(t));
         } else if let Ok(tuple) = idx.downcast::<pyo3::types::PyTuple>(py) {
             // Handle multiple indices e.g. tensor[0,0] or tensor[0:1,0:1]
 
-            if tuple.len() > dims.len() {
+            let mut not_none_count = 0;
```
Comment (LaurentMazare) [inline candle-pyo3/src/lib.rs:492]: Using `tuple.iter.sum` would likely be simpler.

File: candle-pyo3/src/lib.rs [hunk 2]
```diff
@@ -475,27 +477,48 @@ impl PyTensor {
             // Handle a single slice e.g. tensor[0:1] or tensor[0:-1]
             let index = slice.indices(dims[0] as c_long)?;
             indexers.push(Indexer::Slice(index.start as usize, index.stop as usize));
+        } else if let Ok(tensor) = idx.extract::<PyTensor>(py) {
```
[the remark this answers is not in the document]
Comment (LLukas22) [reply candle-pyo3/src/lib.rs:480]: Yup, i missed that case. Should we also support indexing via tensors/lists outside of the first dimension e.g. `t[:,[0,2,4]]`?

[the remark this answers is not in the document]
Comment (LaurentMazare) [reply candle-pyo3/src/lib.rs:480]: Shouldn't that work out of the box? I would have thought that the only limitation is that the indexing tensor/list can only have a single dimension.

[the remark this answers is not in the document]
Comment (LLukas22) [reply candle-pyo3/src/lib.rs:480]: Should probably work, but i will add some tests while i add the list indexing 👍

**Comment by LLukas22** [comment]
Alright should be good to go, i added the list indexing and made sure that it works outside of the first tensor dimension.

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Thanks!

**Pull request merged**